ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs

```diff
@@ -0,0 +1,350 @@
+                let status = validate version fantomasDirectory
+                versionsData.Add(toolVersion, { Version = toolVersion, version; Path = fantomasDirectory; Status = status })
+        | _ ->
+            versionsData.Remove(toolVersion) |> ignore
+            notificationsFired.Remove(toolVersion) |> ignore
+    
+    let invalidateDotnetTools (cache: ToolManifestCache) selectedVersion =
+        cache.LocalToolManifestInfo.Tools
+        |> Seq.tryFind (fun x -> x.Info.PackageId = fantomasPackageId || x.Info.PackageId = fantomasToolPackageId)
+        |> Option.map (fun x -> x.Info.Version, x.PathToExecutable)
+        |> invalidateDotnetTool SolutionDotnetTool
+
+        cache.GlobalToolManifests
+        |> Seq.tryFind (fun x -> (x.Name = fantomasPackageId || x.Name = fantomasToolPackageId) && x.Versions.Count > 0)
+        |> Option.map (fun x ->
+            let version = (Seq.max x.Versions).OriginalVersion
+            version, VirtualFileSystemPath.Parse(calculateGlobalFantomasToolPath version, InteractionContext.Local))
+        |> invalidateDotnetTool GlobalDotnetTool
+
+        match selectedVersion with
+        | FantomasVersionSettings.SolutionDotnetTool ->
+            if versionsData.ContainsKey(SolutionDotnetTool) then () else
+            versionsData.Add(SolutionDotnetTool, { Version = SolutionDotnetTool, ""; Path = null; Status = SelectedButNotFound })
+
+        | FantomasVersionSettings.GlobalDotnetTool ->
+            if versionsData.ContainsKey(GlobalDotnetTool) then () else
+            versionsData.Add(GlobalDotnetTool, { Version = GlobalDotnetTool, ""; Path = null; Status = SelectedButNotFound })
+
+        | _ -> ()
+
+    do
+        if isNull settingsProvider || isNull dotnetToolsTracker then () else
+
+        settingsProvider.Version.Change.Advise(lifetime, fun x ->
+            if not x.HasNew then () else
+            use _ = rwLock.UsingWriteLock()
+            recalculateState x.New)
+        
+        dotnetToolsTracker.ToolManifestCache.Change.Advise(lifetime, fun x ->
+            if not x.HasNew || isNull x.New then () else
+            use _ = rwLock.UsingWriteLock()
+            let settingsVersion = settingsProvider.Version.Value
+            invalidateDotnetTools x.New settingsVersion
+            recalculateState settingsVersion)
+
+    static member Create(lifetime) =
+        FantomasDetector(lifetime, Unchecked.defaultof<FSharpFantomasSettingsProvider>, Unchecked.defaultof<NuGetDotnetToolsTracker>)
+
+    member x.TryRun(runAction: VirtualFileSystemPath -> unit) =
+        use _ = rwLock.UsingWriteLock()
+        fireNotifications()
+        let { Path = path } as versionToRun = versionToRun.Value
+        try runAction path
+        with _ ->
+            versionToRun.Status <- FailedToRun
+            recalculateState settingsProvider.Version.Value
+            x.TryRun(runAction)
+
+    member x.GetSettings() =
+        let settings = Dictionary<_, _>()
```

Comment:
Any reason to specify the wildcard types? It doesn't seem required here. I think it's been asked in previous PRs.